xlators/cluster/dht/src/dht-helper.c

```diff
@@ -1788,22 +1788,39 @@ dht_rebalance_in_progress_check(xlator_t *this, call_frame_t *frame)
 
 int
 dht_inode_ctx_layout_set(inode_t *inode, xlator_t *this,
-                         dht_layout_t *layout_int)
+                         dht_layout_t *layout_int, dht_layout_t **old_layout,
+                         int ref)
 {
     dht_inode_ctx_t *ctx = NULL;
     int ret = -1;
+    uint64_t ctx_int = 0;
 
-    ret = dht_inode_ctx_get(inode, this, &ctx);
-    if (!ret && ctx) {
-        ctx->layout = layout_int;
-    } else {
-        ctx = GF_CALLOC(1, sizeof(*ctx), gf_dht_mt_inode_ctx_t);
-        if (!ctx)
-            return ret;
-        ctx->layout = layout_int;
-    }
+    LOCK(&inode->lock);
+    {
+        ret = __inode_ctx_get(inode, this, &ctx_int);
+        if (!ret) {
+            ctx = (dht_inode_ctx_t *)(uintptr_t)ctx_int;
+            if (ctx) {
+                if (ctx->layout && old_layout)
+                    (*old_layout) = ctx->layout;
+                ctx->layout = layout_int;
+            }
+        }
 
-    ret = dht_inode_ctx_set(inode, this, ctx);
+        if (!ctx) {
+            ctx = GF_CALLOC(1, sizeof(*ctx), gf_dht_mt_inode_ctx_t);
+            if (ctx) {
```

Comment:
If `ctx` is `NULL`, `ret` should be `-1`, but there's a chance that it could be `0`.

Reply:
done

Reply:
As we have already discussed on chat In case of dht_layout_preset reference is not required because we don't take (ref|unref) for preset layout and preset layout has been populated during dht_init only.